src/libsync/filetags.cpp

```diff
@@ -0,0 +1,488 @@
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#ifdef __APPLE__
+#include <CoreFoundation/CoreFoundation.h>
+#endif
+
+#ifdef __linux__
+#include <sys/xattr.h>
+#endif
+
+#include <QStringList>
+#include <QXmlStreamReader>
+
+#include "common/syncjournalfilerecord.h"
+#include "discoveryphase.h"
+#include "filetags.h"
+
+namespace OCC{
+
+Q_LOGGING_CATEGORY(lcFTM, "nextcloud.sync.filetags", QtInfoMsg)
+
+FileTagManager* gFileTagManger=NULL;
+	
+FileTagManager::FileTagManager():QObject(NULL)
+{}
+
+FileTagManager* FileTagManager::GetInstance()
+{
+    if(gFileTagManger==NULL)gFileTagManger=new FileTagManager();
+    return gFileTagManger;
+}
+
+void FileTagManager::fromPropertiesToTagList(QByteArray &list,const QString &properties)
+{
+    if(properties==NULL || properties.isEmpty()) return;
+
+    QStringList tags;
+    QString token;
+    QXmlStreamReader reader("<system-tags>"%properties%"</system-tags>");
+    bool insideTag = false;
+
+    while (!reader.atEnd()){
+        QXmlStreamReader::TokenType type = reader.readNext();
+        QString name = reader.name().toString();
+
+        // Start elements with DAV:
+        if (type == QXmlStreamReader::StartElement && (name=="system-tag" || name=="tag")){
+            insideTag=true;
+        }
+        else if(type== QXmlStreamReader::Characters && insideTag){
+            token.append(reader.text());
+        }
+        else if (type == QXmlStreamReader::EndElement && (name == "system-tag" || name=="tag")){
+            if(token.size()>0)tags << token;
+            token.clear();
+            insideTag=false;
+        }
+    }
+
+    if(tags.size()>0){
+        if(list.size()>0)tags.append(QString(list).split(QChar(0x000A)));
+        tags.removeDuplicates();
+        tags.sort(Qt::CaseInsensitive);
+        list=tags.join(QChar(0x000A)).toUtf8();
+    }
+}
+
+QByteArray FileTagManager::readTagListFromLocalFile(const QString &path)
+{
+#ifdef __APPLE__	
+    // Create necessary system related objects
+    CFStringRef cfstr = path.toCFString();
+    CFURLRef urlref = CFURLCreateWithFileSystemPath(kCFAllocatorDefault,
+                                                    cfstr,
+                                                    kCFURLPOSIXPathStyle,
+                                                    false);
+
+    // Query tags
+    CFArrayRef labels=NULL;
+    Boolean result = CFURLCopyResourcePropertyForKey(urlref,
+                                                     kCFURLTagNamesKey,
+                                                     &labels,
+                                                     NULL);
+
+    QStringList tagarray;
+
+    if(result==true && labels != NULL){
+        // Extract the labels to our array
+        int count = (int) CFArrayGetCount(labels);
+
+        if(count>0){
+            for(int index=0;index<count;index++){
+                CFStringRef str = (CFStringRef)CFArrayGetValueAtIndex(labels,index);
```

Comment:
```suggestion
                const CFStringRef str = (CFStringRef)CFArrayGetValueAtIndex(labels, index);
```

Also would be good to use c++ casts